R/util_data_summaries.R

```diff
@@ -0,0 +1,128 @@
+###############################################################
+###
+###     util_data_summaries.R
+###     Used to summarize data.
+###
+###     To summarize individual variable, call:
+###
+###     ds.summarize_by_column(    data , 
+###                     [func_list] , 
+###                     [codebook] 
+###                 )     
+###
+###     It is anticipated that other data summary functions
+###     will be added later.
+###
+###     Depends on util.R
+###
+###############################################################
+
+# wrap ds.helper functions in an object to protect the namespace
+ds.helper <- list(
+    
+    mean = function(x){
+        if( util.is_vector_of_numbers(x) ){
+            return(mean(as.numeric(x), na.rm=TRUE))
+        }else{
+            return(NA)
+        }
+    }
+    ,
+    pct_blank = function(x){
+        mean(util.is_blank(x)) * 100
+    }
+    ,
+    sd = function(x){
+        sd(x, na.rm=TRUE)
+    }
+    ,
+    obs_min = function(x){
+        if( util.is_vector_of_numbers(x) ){
+            return(min(as.numeric(x),na.rm=T))
+        }else{
+            return(NA)   
+        }
+    }
+    ,
+    obs_max = function(x){
+        if( util.is_vector_of_numbers(x) ){
+            return(max(as.numeric(x),na.rm=T))
+        }else{
+            return(NA)   
+        }
+    }
+    ,
+    logical_to_numeric = function(x){
+        # if a value is logical (boolean), convert to numeric for calculation
+        if(class(x) %in% "logical"){ x <- as.numeric(x) }
+        return(x)
+    }
+    ,
+    n_unique = function(x){
+        # includes NA
+        length(unique(x))
+    }
+)
+
+# these must be defined outside ds.helper because they reference
+# functions defined in it
```

Comment:
Two alternative patterns for this `ds.helper` definition:

1. "All-dollar":

```
ds.helper <- list()

ds.helper$foo <- function () {}

ds.helper$bar <- function () { ds.helper$foo() + 1 }
```

2. Javascript module pattern; allows private variables.

```
ds.helper <- (function () {
  private_foo <- function () {}

  foo <- function () {}
  bar <- function () { bar() + 1 }

  exports <- list(
    foo = function () {},
    bar = function () {}
  )
})()
```